### PR TITLE
[JENKINS-49994] Warn about Remoting serialization of anonymous inner classes

### PR DESCRIPTION
**Pull request by jglick**
See [JENKINS-49994](https://issues.jenkins-ci.org/browse/JENKINS-49994). Downstream of https://github.com/jenkinsci/remoting/pull/259. Amends #3312 to capture more events. Subsumes #3343.

Also solves a number of particular cases, as was started in #3071 for the benefit of https://github.com/jenkinsci/support-core-plugin/pull/128. I did not attempt to do a code search as in [JENKINS-49987](https://issues.jenkins-ci.org/browse/JENKINS-49987); rather I solved those cases which can be reproduced by running a freestyle or Pipeline build on an agent which runs a shell step and archives an artifact with fingerprinting, along with a few other minor use cases.

### Proposed changelog entries

* Update Remoting from 3.17 to 3.18 in order to apply various performance and diagnosibility improvements ([full changelog](https://github.com/jenkinsci/remoting/blob/master/CHANGELOG.md#318)):
  * [JENKINS-49994](https://issues.jenkins-ci.org/browse/JENKINS-49994): warnings sent to the system and agent logs when anonymous classes are serialized over a Remoting channel
  * [JENKINS-49415](https://issues.jenkins-ci.org/browse/JENKINS-49415): add uncaught exception handler to the `Engine`'s executor service
  * [JENKINS-49472](https://issues.jenkins-ci.org/browse/JENKINS-49472): log channel name in `StreamCorruptedException`s
  * [JENKINS-48561](https://issues.jenkins-ci.org/browse/JENKINS-48561): give precedence to proxy exclusion list system property over environmental vars
  * performance optimization in remote class loading

@reviewbybees @jenkinsci/code-reviewers

### CODE REVIEW

File: core/src/main/java/hudson/FilePath.java [hunk 1]
```diff
@@ -1105,23 +1104,28 @@ protected void after() {}
      * @since 1.522
      */
     public <V> Callable<V,IOException> asCallableWith(final FileCallable<V> task) {
-        return new Callable<V,IOException>() {
-            @Override
-            public V call() throws IOException {
-                try {
-                    return act(task);
```
Comment (jglick) [inline core/src/main/java/hudson/FilePath.java]: [Whitespace-reduced diff](https://github.com/jenkinsci/jenkins/pull/3333/files?w=1) is easier to review.

File: core/src/main/java/hudson/FilePath.java [hunk 2]
```diff
@@ -1105,23 +1104,28 @@ protected void after() {}
...
-                } catch (InterruptedException e) {
-                    throw (IOException)new InterruptedIOException().initCause(e);
-                }
+        return new CallableWith<>(task);
+    }
+    private class CallableWith<V> implements Callable<V, IOException> {
```
Comment (jglick) [inline core/src/main/java/hudson/FilePath.java]: You may be surprised to see a number of these nested classes being non`static`. This was just the simplest refactoring that addresses the essential problems of using anonymous inner classes as described in https://github.com/jenkinsci/jenkins/pull/3312#issuecomment-368713851. You could make the nested classes `static`, but then you would need to define a `FilePath` field anyway, which just makes the code more verbose and does nothing beyond rename the implicit `this$0` field.

So why do all these nested classes inside `FilePath` _need_ that field to be serialized, you might wonder? Well, for its `SoloFilePathFilter filter` field, passed to calls like `reading`, delegating to methods like `FilePathFilter.read`.

OK, so why not define a `SoloFilePathFilter filter` field in a `static` nested class then? Well, because it is not `Serializable`.

Huh? How can it not be serializable, if it is a field we are transferring to the remote side and then using? In fact it is `transient`. And it is restored in `FilePath.readObject` by some complicated logic which takes into account the serial form of `FilePath`.

Basically, this would only be possible if `FilePathFilter` were given a `writeReplace` method which produced some sort of placeholder object based on `Channel.current() == null` which `readResolve`d to either `null` or `SoloFilePathFilter.wrap(FilePathFilter.current())`. A bigger refactoring than I felt comfortable with.

**Review by ghost — APPROVED** [review]

**Review by ghost — DISMISSED** [review]
nm, test failure looks related.  https://ci.jenkins.io/job/Core/job/jenkins/view/change-requests/job/PR-3333/lastCompletedBuild/testReport/hudson.logging/LogRecorderManagerTest/Windows___Windows_Publishing___loggingOnSlaves/

**Review by ghost — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Sorry, it's not JEP-200 and hence I have no excuse to review it

File: pom.xml [hunk 3]
```diff
@@ -104,7 +104,7 @@ THE SOFTWARE.
     <maven-war-plugin.version>3.0.0</maven-war-plugin.version> <!-- JENKINS-47127 bump when 3.2.0 is out. Cf. MWAR-407 -->
 
     <!-- Minimum Remoting version, which is tested for API compatibility -->
-    <remoting.version>3.17</remoting.version>
+    <remoting.version>3.18</remoting.version>
```
Comment (jglick) [inline pom.xml]: [Changes](https://github.com/jenkinsci/remoting/compare/remoting-3.17...remoting-3.18)

File: core/src/main/java/jenkins/slaves/restarter/JnlpSlaveRestarterInstaller.java [hunk 4]
```diff
@@ -34,67 +35,80 @@
...
-                        @Override
-                        public void onReconnect() {
+            // filter out ones that doesn't apply
+            for (Iterator<SlaveRestarter> itr = restarters.iterator(); itr.hasNext(); ) {
+                SlaveRestarter r =  itr.next();
+                if (!r.canWork())
+                    itr.remove();
+            }
+
+            e.addListener(new EngineListenerAdapter() {
```
Comment (timja) [inline core/src/main/java/jenkins/slaves/restarter/JnlpSlaveRestarterInstaller.java:91]: https://github.com/jenkinsci/jenkins/pull/6315
